### PR TITLE
web: Support transparent wmode

### LINKED ISSUE #1752: web: Support transparent wmode
When the flash background is transparent, black is used instead.

### PR DESCRIPTION
**Pull request by Herschel**
Recording 2022-04-13 at 15 40 33

 * Change wgpu/WebGL shaders to output premultiplied alpha.
   * Browser engines themselves work in premultiplied alpha, so output this for proper alpha blending on a transparent canvas.
   * This will help with #6629 because Flash bitmaps are stored with premultiplied alpha.
 * Add "wmode" Flash embed parameter to web config.
   * Currently only "transparent" has an effect.
   * Perhaps we can simulate "window" vs. "opaque" using something like `z-index: 999`.
   * "gpu"/"direct" will always be ignored.

Fixes #1752.

### CODE REVIEW

File: core/src/display_object/stage.rs [hunk 1]
```diff
@@ -1060,3 +1087,67 @@ impl FromWStr for StageQuality {
...
+
+impl Display for WindowMode {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        // Match string values returned by AS.
+        let s = match *self {
+            WindowMode::Window => "window",
+            WindowMode::Opaque => "opaque",
+            WindowMode::Transparent => "transparent",
+            WindowMode::Direct => "direct",
+            WindowMode::Gpu => "direct",
```
Comment (relrelb) [inline core/src/display_object/stage.rs:1133]: `"gpu"`

**Comment by Toad06** [comment]
>       * Perhaps we can simulate "window" vs. "opaque" using something like `z-index: 999`.

I believe that https://github.com/ruffle-rs/ruffle/issues/5344#issuecomment-948065397 could actually need something like this (though the issue itself is not about wmode). Fun fact: On Basilisk, adding a wmode attribute to the embed tag won't hide an overlayed text, unless the attribute is specified with... a wrong value.

**Comment by Herschel** [comment]
> I believe that [#5344 (comment)](https://github.com/ruffle-rs/ruffle/issues/5344#issuecomment-948065397) could actually need something like this (though the issue itself is not about wmode). Fun fact: On Basilisk, adding a wmode attribute to the embed tag won't hide an overlayed text, unless the attribute is specified with... a wrong value.

The [Flash docs](https://helpx.adobe.com/flash/kb/flash-object-embed-tag-attributes.html) say that `wmode="window"` lets the "browser determines how the SWF content is layered ... you cannot explicitly specify if SWF content appears above or below other HTML elements." So I guess it's actually not specified what the behavior is.

@Toad06 What do you think? Do you think we should try to render "window" always-on-top?

In Waterfox Classic at least, "window" seems to act the same as "opaque", and HTML renders on top of the Flash embed as normal. "gpu" is the only one that seems to always render on top.

**Comment by Toad06** [comment]
I'd say to keep the `wmode="window"` behavior as it is now, since it can vary from browser to browser - at least this is what I understand from the docs (thanks for sharing the link 😉). I wanted to check what was the behavior on Internet Explorer using my old Win7 PC to make a comparison, but the plugin got deleted...

**Comment by Toad06** [comment]
It also seems like "gpu" (or a wrong value) doesn't always render on top. I found this while testing the shadowbox.js library.

[Here is the sample I made.](https://github.com/ruffle-rs/ruffle/files/8501859/_sample.zip)

**Comment by Herschel** [comment]
Let's not bother fiddling with the z-index for now -- the user could always do it themselves, or maybe this could be a separate option someday. It's fairly easy to implement if we ever want to, doing something like `:host([wmode="window"]) { z-index: 999; }`. Thanks!

**Pull request merged**